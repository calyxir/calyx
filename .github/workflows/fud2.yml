name: fud2 tests

on:
  push:
  pull_request:
    branches: [main]

jobs:
  test:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v4
      - name: Install Rust
        uses: actions-rust-lang/setup-rust-toolchain@v1
      - run: cargo test -p fud
