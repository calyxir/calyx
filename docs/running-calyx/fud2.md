# fud2: An Experimental Successor to fud

[fud][] is the compiler driver tool for orchestrating the Calyx ecosystem.
fud2 is an experiment in building a new driver that works like fud that adds some fundamental new capabilities and resolves some underlying problems.

"Original" fud is still the right tool for almost all jobs; fud2 is in an experimental phase and does not support everything fud can do.
Someday, fud2 may supplant fud, but it needs more work before it is ready to do that.
Until then, fud remains your first choice for all your build-related needs.

[fud]: ./fud/index.md

## Set Up

fud2 is a Rust tool, so you can build it along with everything else in this monorepo with `cargo build`.
You might then want to do something like ``ln -s `pwd`/target/debug/fud2 ~/.local/bin`` for easy access to the `fud2` binary.

fud2 depends on [Ninja][].
Install it using your OS package manager or by downloading a binary.

### Configuration

Run the following command to edit `fud2`'s configuration file (usually `~/.config/fud2.toml`):

    $ fud2 edit-config

Add these lines:

```toml
[calyx]
base = "<path to calyx checkout>"
```

Now you're ready to use fud2.

[ninja]: https://ninja-build.org

## General Use

You can see complete command-line documentation with `fud2 --help`.
But generally, you want to do something like this:

    $ fud2 <IN> -o <OUT>

For example, use this to compile a Calyx program to Verilog:

    $ fud2 foo.futil -o bar.sv

fud2 tries to automatically guess the input and output formats using filename extensions.
If that doesn't work, you can choose for it with `--from <state>` and `--to <state>`;
for example, this is a more explicit version of the above:

    $ fud2 foo.futil -o bar.sv --from calyx --to verilog

You can also omit the input and output filenames to instead use stdin and stdout.
In that case, `--from` and `--to` respectively are required.
So here's yet another way to do the same thing:

    $ fud2 --from calyx --to verilog < foo.futil > bar.sv

This is handy if you just want to print the result of a build to the console:

    $ fud2 foo.futil --to verilog

Some operations use other configuration options, which can come from either your `fud2.toml` or the command line.
Use `--set key=value` to override any such option.

## Advanced Options

Use `fud2 --help` for an overview of the command-line interface.
Here are some options you might need:

* By default, fud2 runs the build in a directory called `.fud2` within the working directory. It automatically deletes this directory when the build is done.
    * It can be useful to keep this build directory around for debugging or as a "cache" for future builds. Use `--keep` to prevent fud2 from deleting the build directory.
    * You can also tell fud2 to use a different build directory with `--dir`. If you give it an existing directory, it will never be deleted, even without `--keep`. (Only "fresh" build directories are automatically cleaned up.)
* If you don't like the operation path that fud2 selected for your build, you can control it with `--through <OP>`. fud2 will search the operation graph for a path that contains that op. You can provide this option multiple times; fud2 will look for paths that contain *all* these operations, in order.
* You can choose one of several modes with `-m <NAME>`:
    * `run`: Actually execute a build. The default.
    * `gen`: Generate the Ninja build file in the build directory, but don't actually run the build. The default `run` mode is therefore approximately like doing `fud2 -m gen && ninja -C .fud2`.
    * `emit`: Just print the Ninja build file to stdout. The `gen` mode is therefore approximately `fud2 -m emit > .fud2/build.ninja`.
    * `plan`: Print a brief description of the plan, i.e., the sequence of operations that the build would run.
    * `dot`: Print a [GraphViz][] depiction of the plan. Try `fud2 -m dot | dot -Tpdf > graph.pdf` and take a look.

There are also some subcommands for doing things other than building stuff:

* `fud2 edit-config`: Open the fud2 configuration file in `$EDITOR`.
* `fud2 list`: Print out all the available states and operations.
* `fud2 get-rsrc FILE`: Fetch a *resource file* and place it in the working directory. You typically do not need to use this interactively; it is used during builds to obtain files included with fud2 that are necessary for a given build.

[graphviz]: https://graphviz.org

## The Design of fud2

<div class="warning">

This section is about the *implementation* of fud2; it is only relevant if you want to work on it yourself.
No need to read any farther if all you want is to *use* fud2.

</div>

### fud2 is a Command Orchestrator

fud2 consists of two pieces, which are two separate Rust crates:

* FudCore (the `fud-core` crate): is a *generic compiler driver* library. This library is not specific to Calyx and could hypothetically be used to build a fud2-like driver for any compiler ecosystem. Clients of the `fud-core` library work by constructing a `Driver` object that encapsulates a set of *states* and *operations* that define the driver's behavior.
* fud2 itself is a program that uses the FudCore library. All of the Calyx-specific logic lives in `fud2`. For the most part, all of the code in the `fud2` crate consists of declaring a bunch of states and operations. The `main` function does little more than dispatch to the resulting `Driver` object's generic command-line interface.

The central design philosophy of FudCore (and by extension, fud2 itself) is that its sole job is to orchestrate external functionality.
All that functionality must be available as separate tools that can be invoked via the command line.
This is an important goal because it means the driver has a clear, discrete goal: *all it does* is decide on a list of commands to execute to perform a build.
All the "interesting work" must be delegated to separate tools outside of the driver.
This philosophy has both advantages and disadvantages:

* On the positive side, it forces all the interesting logic to be invokable via a command that you, the user, can run equally well yourself. So if something is going wrong, there is *always* a command line you can copy and paste into your terminal to reproduce the problem at that particular step. It also means that the input and output of every step must be written to files in the filesystem, so you can easily inspect the intermediate state between every command. This file-based operation also means that fud2 builds are parallel and incremental by default.
* On the other hand, requiring everything to be separate commands means that fud2 has a complicated dependency story. It is not a monolith: to get meaningful work done, you currently have to install a bunch of Python components (among other things) so fud2 can invoke them. (We hope to mitigate the logistical pain this incurs over time, but we're not there yet.) Also, writing everything to a file in between each step comes at a performance cost. Someday, it may be a performance bottleneck that two steps in a build cannot simply exchange their data directly, through memory, and must serialize everything to disk first. (This has not been a problem in practice yet.)

If you want to extend fud2 to do something new, the consequence is that you first need to come up with a sequence of commands that do that thing.
If necessary, you may find that you need to create new executables to do some minor glue tasks that would otherwise be implicit.
Then "all you need to do" is teach fud2 to execute those commands.

### States, Operations, and Setups

You can think of a FudCore driver as a graph, where the vertices are *states* and the edges are *operations*.
(In fact, this is literally the graph you can visualize with `-m dot`.)
Any build is a transformation from one state to another, traversing a path through this graph.
The operations (edges) along this path are the commands that must be executed to transform a file from the initial state to the final state.

To make fud2 do something new, you probably want to add one or more operations, and you may need to add new states.
Aside from declaring the source and destination states,
operations generate chunks of [Ninja][] code.
So to implement an operation, you write a Rust function with this signature:

    fn build(emitter: &mut Emitter, input: &str, output: &str)

Here, `emitter` is a wrapper around an output stream with a bunch of utility functions for printing out lines of Ninja code.
`input` and `output` are filenames.
So your job in this function is to print (at least) a Ninja `build` command that produces `output` as a target and uses `input` as a dependency.
For example, the Calyx-to-Verilog compiler operation might emit this chunk of Ninja code:

    build bar.sv: calyx foo.futil
      backend = verilog

when the `input` argument above is `"foo.futil"` and the `output` is `"bar.sv"`.
(The FudCore library will conjure these filenames for you; your job in this operation is just to use them as is.)

Notice here that the generated Ninja chunk is using a build rule called `calyx`.
This also needs to be defined.
To set up things like variables and build rules that operations can use, FudCore has a separate concept called *setups*.
A setup is a function that generates some Ninja code that might be shared among multiple operations (or multiple instances of the same operation).
For example, our setup for Calyx compilation generates code like this:

    calyx-base = /path/to/calyx
    calyx-exe = $calyx-base/target/debug/calyx
    rule calyx
      command = $calyx-exe -l $calyx-base -b $backend $args $in > $out

That is, it defines two Ninja variables and one Ninja ruleâ€”the one that our build command above uses.
We *could* have designed FudCore without a separation between setups and operations, so this rule would get declared right next to the `build` command above.
But that would end up duplicating a lot of setup code that really only needs to appear once.
So that's why setups exist: to share a single stanza of Ninja scaffolding code between multiple operations.
