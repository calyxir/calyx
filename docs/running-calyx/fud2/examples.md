# Examples

These commands will assume you're in the root directory of the Calyx
repository and that you have compiled Calyx.

**Simulating Calyx.**
Fud2 can compile a Calyx program to Verilog and simulate it using Verilator.

```bash
# Compile and simulate a vectorized add implementation
# in Calyx using the data provided,
# then dump the vcd into a new file for debugging.
# === Calyx:   examples/futil/vectorized-add.futil
# === data:    examples/dahlia/vectorized-add.fuse.data
# === output:  v-add.vcd
fud2 examples/futil/vectorized-add.futil \
  -o v-add.vcd \
  -s sim.data=examples/dahlia/vectorized-add.fuse.data \
```

**Simulating Dahlia.**
The following command prints out the final state of all memories by specifying
`--to dat`.

```bash
# Compile a Dahlia dot product implementation and
# simulate in verilog using the data provided.
# === Dahlia: examples/dahlia/dot-product.fuse
# === data:   examples/dahlia/dot-product.fuse.data
#     (`.data` is used as an extension alias for `.json`)
fud2 examples/dahlia/dot-product.fuse \
  --to dat \
  -s sim.data=examples/dahlia/dot-product.fuse.data
```

**Interpreting Calyx.**
In addition to Verilator, fud can execute Calyx programs using the [interpreter](../interpreter.md).

```bash
# Execute a Calyx program without compiling it,
# producing a JSON snapshot of the final state.
# === Calyx:   tests/correctness/while.futil
# === data:    tests/correctness/while.futil.data
fud2 tests/correctness/while.futil \
  --to dat \
  -s sim.data=tests/correctness/while.futil.data \
  --through cider
```