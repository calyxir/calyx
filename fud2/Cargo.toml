[package]
name = "fud"
version = "0.0.2"
edition.workspace = true
authors.workspace = true
license-file.workspace = true
repository.workspace = true
homepage.workspace = true

keywords = ["build-tool"]
readme = "README.md"
categories = ["build-tool"]
description = "Compiler driver for the Calyx infrastructure"

[dependencies]
fud-core = { path = "fud-core", version = "0.0.2" }
anyhow.workspace = true
manifest-dir-macros = "0.1"
include_dir = "0.7"

[lib]
name = "fud2"
path = "src/lib.rs"

[[bin]]
name = "fud2"
path = "src/main.rs"

[dev-dependencies]
insta = "1.36.0"
