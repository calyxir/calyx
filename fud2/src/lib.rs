use fud_core::{
    exec::{SetupRef, StateRef},
    run::{EmitResult, Emitter},
    DriverBuilder,
};

fn setup_calyx(
    bld: &mut DriverBuilder,
    verilog: StateRef,
) -> (StateRef, SetupRef) {
    let calyx = bld.state("calyx", &["futil"]);
    let calyx_setup = bld.setup("Calyx compiler", |e| {
        e.config_var("calyx-base", "calyx.base")?;
        e.config_var_or(
            "calyx-exe",
            "calyx.exe",
            "$calyx-base/target/debug/calyx",
        )?;
        e.rule(
            "calyx",
            "$calyx-exe -l $calyx-base -b $backend $args $in > $out",
        )?;
        e.rule(
            "calyx-pass",
            "$calyx-exe -l $calyx-base -p $pass $args $in > $out",
        )?;
        Ok(())
    });
    bld.op(
        "calyx-to-verilog",
        &[calyx_setup],
        calyx,
        verilog,
        |e, input, output| {
            e.build_cmd(&[output], "calyx", &[input], &[])?;
            e.arg("backend", "verilog")?;
            Ok(())
        },
    );
    (calyx, calyx_setup)
}

fn setup_dahlia(
    bld: &mut DriverBuilder,
    calyx: StateRef,
) -> (StateRef, SetupRef) {
    let dahlia = bld.state("dahlia", &["fuse"]);
    let dahlia_setup = bld.setup("Dahlia compiler", |e| {
        e.config_var("dahlia-exe", "dahlia")?;
        e.rule(
            "dahlia-to-calyx",
            "$dahlia-exe -b calyx --lower -l error $in -o $out",
        )?;
        Ok(())
    });
    bld.rule(&[dahlia_setup], dahlia, calyx, "dahlia-to-calyx");
    (dahlia, dahlia_setup)
}

fn setup_mrxl(
    bld: &mut DriverBuilder,
    calyx: StateRef,
) -> (StateRef, SetupRef) {
    let mrxl = bld.state("mrxl", &["mrxl"]);
    let mrxl_setup = bld.setup("MrXL compiler", |e| {
        e.var("mrxl-exe", "mrxl")?;
        e.rule("mrxl-to-calyx", "$mrxl-exe $in > $out")?;
        Ok(())
    });
    bld.rule(&[mrxl_setup], mrxl, calyx, "mrxl-to-calyx");
    (mrxl, mrxl_setup)
}

pub fn build_driver(bld: &mut DriverBuilder) {
    // The verilog state
    let verilog = bld.state("verilog", &["sv", "v"]);
    // Calyx.
    let (calyx, calyx_setup) = setup_calyx(bld, verilog);
    // Dahlia.
    setup_dahlia(bld, calyx);
    // MrXL.
    setup_mrxl(bld, calyx);

    // Shared machinery for RTL simulators.
    let dat = bld.state("dat", &["json"]);
    let vcd = bld.state("vcd", &["vcd"]);
    let simulator = bld.state("sim", &["exe"]);
    let sim_setup = bld.setup("RTL simulation", |e| {
        // Data conversion to and from JSON.
        e.config_var_or("python", "python", "python3")?;
        e.rsrc("json-dat.py")?;
        e.rule("hex-data", "$python json-dat.py --from-json $in $out")?;
        e.rule("json-data", "$python json-dat.py --to-json $out $in")?;

        // The Verilog testbench.
        e.rsrc("tb.sv")?;

        // The input data file. `sim.data` is required.
        let data_name = e.config_val("sim.data")?;
        let data_path = e.external_path(data_name.as_ref());
        e.var("sim_data", data_path.as_str())?;

        // Produce the data directory.
        e.var("datadir", "sim_data")?;
        e.build_cmd(
            &["$datadir"],
            "hex-data",
            &["$sim_data"],
            &["json-dat.py"],
        )?;

        // Rule for simulation execution.
        e.rule(
            "sim-run",
            "./$bin +DATA=$datadir +CYCLE_LIMIT=$cycle-limit $args > $out",
        )?;

        // More shared configuration.
        e.config_var_or("cycle-limit", "sim.cycle_limit", "500000000")?;

        Ok(())
    });
    bld.op(
        "simulate",
        &[sim_setup],
        simulator,
        dat,
        |e, input, output| {
            e.build_cmd(&["sim.log"], "sim-run", &[input, "$datadir"], &[])?;
            e.arg("bin", input)?;
            e.arg("args", "+NOTRACE=1")?;
            e.build_cmd(
                &[output],
                "json-data",
                &["$datadir", "sim.log"],
                &["json-dat.py"],
            )?;
            Ok(())
        },
    );
    bld.op("trace", &[sim_setup], simulator, vcd, |e, input, output| {
        e.build_cmd(
            &["sim.log", output],
            "sim-run",
            &[input, "$datadir"],
            &[],
        )?;
        e.arg("bin", input)?;
        e.arg("args", &format!("+NOTRACE=0 +OUT={}", output))?;
        Ok(())
    });

    // Icarus Verilog.
    let verilog_noverify = bld.state("verilog-noverify", &["sv"]);
    let icarus_setup = bld.setup("Icarus Verilog", |e| {
        e.var("iverilog", "iverilog")?;
        e.rule("icarus-compile", "$iverilog -g2012 -o $out tb.sv $in")?;
        Ok(())
    });
    bld.op(
        "calyx-noverify",
        &[calyx_setup],
        calyx,
        verilog_noverify,
        |e, input, output| {
            // Icarus requires a special --disable-verify version of Calyx code.
            e.build_cmd(&[output], "calyx", &[input], &[])?;
            e.arg("backend", "verilog")?;
            e.arg("args", "--disable-verify")?;
            Ok(())
        },
    );
    bld.op(
        "icarus",
        &[sim_setup, icarus_setup],
        verilog_noverify,
        simulator,
        |e, input, output| {
            e.build_cmd(&[output], "icarus-compile", &[input], &["tb.sv"])?;
            Ok(())
        },
    );

    // Calyx to FIRRTL.
    let firrtl = bld.state("firrtl", &["fir"]);
    bld.op(
        "calyx-to-firrtl",
        &[calyx_setup],
        calyx,
        firrtl,
        |e, input, output| {
            e.build_cmd(&[output], "calyx", &[input], &[])?;
            e.arg("backend", "firrtl")?;
            Ok(())
        },
    );

    // The FIRRTL compiler.
    let firrtl_setup = bld.setup("Firrtl to Verilog compiler", |e| {
        e.config_var("firrtl-exe", "firrtl.exe")?;
        e.rule("firrtl", "$firrtl-exe -i $in -o $out -X sverilog")?;

        e.rsrc("primitives-for-firrtl.sv")?;
        e.rule(
            "add-firrtl-prims",
            "cat primitives-for-firrtl.sv $in > $out",
        )?;

        Ok(())
    });
    fn firrtl_compile(
        e: &mut Emitter,
        input: &str,
        output: &str,
    ) -> EmitResult {
        let tmp_verilog = "partial.sv";
        e.build_cmd(&[tmp_verilog], "firrtl", &[input], &[])?;
        e.build_cmd(
            &[output],
            "add-firrtl-prims",
            &[tmp_verilog],
            &["primitives-for-firrtl.sv"],
        )?;
        Ok(())
    }
    bld.op("firrtl", &[firrtl_setup], firrtl, verilog, firrtl_compile);
    // This is a bit of a hack, but the Icarus-friendly "noverify" state is identical for this path
    // (since FIRRTL compilation doesn't come with verification).
    bld.op(
        "firrtl-noverify",
        &[firrtl_setup],
        firrtl,
        verilog_noverify,
        firrtl_compile,
    );

    // primitive-uses backend
    let primitive_uses_json = bld.state("primitive-uses-json", &["json"]);
    bld.op(
        "primitive-uses",
        &[calyx_setup],
        calyx,
        primitive_uses_json,
        |e, input, output| {
            e.build_cmd(&[output], "calyx", &[input], &[])?;
            e.arg("backend", "primitive-uses")?;
            Ok(())
        },
    );

    // Verilator.
    let verilator_setup = bld.setup("Verilator", |e| {
        e.config_var_or("verilator", "verilator.exe", "verilator")?;
        e.config_var_or("cycle-limit", "sim.cycle_limit", "500000000")?;
        e.rule(
            "verilator-compile",
            "$verilator $in tb.sv --trace --binary --top-module TOP -fno-inline -Mdir $out-dir",
        )?;
        e.rule("cp", "cp $in $out")?;
        Ok(())
    });
    bld.op(
        "verilator",
        &[sim_setup, verilator_setup],
        verilog,
        simulator,
        |e, input, output| {
            let out_dir = "verilator-out";
            let sim_bin = format!("{}/VTOP", out_dir);
            e.build_cmd(
                &[&sim_bin],
                "verilator-compile",
                &[input],
                &["tb.sv"],
            )?;
            e.arg("out-dir", out_dir)?;
            e.build("cp", &sim_bin, output)?;
            Ok(())
        },
    );

    // Interpreter.
    let debug = bld.state("debug", &[]); // A pseudo-state.
    let cider_setup = bld.setup("Cider interpreter", |e| {
        e.config_var_or(
            "cider-exe",
            "cider.exe",
            "$calyx-base/target/debug/cider",
        )?;
        e.rule(
            "cider",
            "$cider-exe -l $calyx-base --raw --data data.json $in > $out",
        )?;
        e.rule(
            "cider-debug",
            "$cider-exe -l $calyx-base --data data.json $in debug || true",
        )?;
        e.arg("pool", "console")?;

        // TODO Can we reduce the duplication around and `$python`?
        e.rsrc("interp-dat.py")?;
        e.config_var_or("python", "python", "python3")?;
        e.rule("dat-to-interp", "$python interp-dat.py --to-interp $in")?;
        e.rule(
            "interp-to-dat",
            "$python interp-dat.py --from-interp $in $sim_data > $out",
        )?;
        e.build_cmd(
            &["data.json"],
            "dat-to-interp",
            &["$sim_data"],
            &["interp-dat.py"],
        )?;
        Ok(())
    });
    bld.op(
        "interp",
        &[sim_setup, calyx_setup, cider_setup],
        calyx,
        dat,
        |e, input, output| {
            let out_file = "interp_out.json";
            e.build_cmd(&[out_file], "cider", &[input], &["data.json"])?;
            e.build_cmd(
                &[output],
                "interp-to-dat",
                &[out_file],
                &["$sim_data", "interp-dat.py"],
            )?;
            Ok(())
        },
    );
    bld.op(
        "debug",
        &[sim_setup, calyx_setup, cider_setup],
        calyx,
        debug,
        |e, input, output| {
            e.build_cmd(&[output], "cider-debug", &[input], &["data.json"])?;
            Ok(())
        },
    );

    // Xilinx compilation.
    let xo = bld.state("xo", &["xo"]);
    let xclbin = bld.state("xclbin", &["xclbin"]);
    let xilinx_setup = bld.setup("Xilinx tools", |e| {
        // Locations for Vivado and Vitis installations.
        e.config_var("vivado-dir", "xilinx.vivado")?;
        e.config_var("vitis-dir", "xilinx.vitis")?;

        // Package a Verilog program as an `.xo` file.
        e.rsrc("gen_xo.tcl")?;
        e.rsrc("get-ports.py")?;
        e.config_var_or("python", "python", "python3")?;
        e.rule("gen-xo", "$vivado-dir/bin/vivado -mode batch -source gen_xo.tcl -tclargs $out `$python get-ports.py kernel.xml`")?;
        e.arg("pool", "console")?;  // Lets Ninja stream the tool output "live."

        // Compile an `.xo` file to an `.xclbin` file, which is where the actual EDA work occurs.
        e.config_var_or("xilinx-mode", "xilinx.mode", "hw_emu")?;
        e.config_var_or("platform", "xilinx.device", "xilinx_u50_gen3x16_xdma_201920_3")?;
        e.rule("compile-xclbin", "$vitis-dir/bin/v++ -g -t $xilinx-mode --platform $platform --save-temps --profile.data all:all:all --profile.exec all:all:all -lo $out $in")?;
        e.arg("pool", "console")?;

        Ok(())
    });
    bld.op(
        "xo",
        &[calyx_setup, xilinx_setup],
        calyx,
        xo,
        |e, input, output| {
            // Emit the Verilog itself in "synthesis mode."
            e.build_cmd(&["main.sv"], "calyx", &[input], &[])?;
            e.arg("backend", "verilog")?;
            e.arg("args", "--synthesis -p external")?;

            // Extra ingredients for the `.xo` package.
            e.build_cmd(&["toplevel.v"], "calyx", &[input], &[])?;
            e.arg("backend", "xilinx")?;
            e.build_cmd(&["kernel.xml"], "calyx", &[input], &[])?;
            e.arg("backend", "xilinx-xml")?;

            // Package the `.xo`.
            e.build_cmd(
                &[output],
                "gen-xo",
                &[],
                &[
                    "main.sv",
                    "toplevel.v",
                    "kernel.xml",
                    "gen_xo.tcl",
                    "get-ports.py",
                ],
            )?;
            Ok(())
        },
    );
    bld.op("xclbin", &[xilinx_setup], xo, xclbin, |e, input, output| {
        e.build_cmd(&[output], "compile-xclbin", &[input], &[])?;
        Ok(())
    });

    // Xilinx execution.
    // TODO Only does `hw_emu` for now...
    let xrt_setup = bld.setup("Xilinx execution via XRT", |e| {
        // Generate `emconfig.json`.
        e.rule("emconfig", "$vitis-dir/bin/emconfigutil --platform $platform")?;
        e.build_cmd(&["emconfig.json"], "emconfig", &[], &[])?;

        // Execute via the `xclrun` tool.
        e.config_var("xrt-dir", "xilinx.xrt")?;
        e.rule("xclrun", "bash -c 'source $vitis-dir/settings64.sh ; source $xrt-dir/setup.sh ; XRT_INI_PATH=$xrt_ini EMCONFIG_PATH=. XCL_EMULATION_MODE=$xilinx-mode $python -m fud.xclrun --out $out $in'")?;
        e.arg("pool", "console")?;

        // "Pre-sim" and "post-sim" scripts for simulation.
        e.rule("echo", "echo $contents > $out")?;
        e.build_cmd(&["pre_sim.tcl"], "echo", &[""], &[""])?;
        e.arg("contents", "open_vcd\\\\nlog_vcd *\\\\n")?;
        e.build_cmd(&["post_sim.tcl"], "echo", &[""], &[""])?;
        e.arg("contents", "close_vcd\\\\n")?;

        Ok(())
    });
    bld.op(
        "xrt",
        &[xilinx_setup, sim_setup, xrt_setup],
        xclbin,
        dat,
        |e, input, output| {
            e.rsrc("xrt.ini")?;
            e.build_cmd(
                &[output],
                "xclrun",
                &[input, "$sim_data"],
                &["emconfig.json", "xrt.ini"],
            )?;
            e.arg("xrt_ini", "xrt.ini")?;
            Ok(())
        },
    );
    bld.op(
        "xrt-trace",
        &[xilinx_setup, sim_setup, xrt_setup],
        xclbin,
        vcd,
        |e, input, output| {
            e.rsrc("xrt_trace.ini")?;
            e.build_cmd(
                &[output], // TODO not the VCD, yet...
                "xclrun",
                &[input, "$sim_data"],
                &[
                    "emconfig.json",
                    "pre_sim.tcl",
                    "post_sim.tcl",
                    "xrt_trace.ini",
                ],
            )?;
            e.arg("xrt_ini", "xrt_trace.ini")?;
            Ok(())
        },
    );

    let yxi = bld.state("yxi", &["yxi"]);
    bld.op(
        "calyx-to-yxi",
        &[calyx_setup],
        calyx,
        yxi,
        |e, input, output| {
            e.build_cmd(&[output], "calyx", &[input], &[])?;
            e.arg("backend", "yxi")?;
            Ok(())
        },
    );

    let wrapper_setup = bld.setup("YXI and AXI generation", |e| {
        // Define a `gen-axi` rule that invokes our Python code generator program.
        // For now point to standalone axi-generator.py. Can maybe turn this into a rsrc file?
        e.config_var_or(
            "axi-generator",
            "axi.generator",
            "$calyx-base/yxi/axi-calyx/axi-generator.py",
        )?;
        e.config_var_or("python", "python", "python3")?;
        e.rule("gen-axi", "$python $axi-generator $in > $out")?;

        // Define a simple `combine` rule that just concatenates any numer of files.
        e.rule("combine", "cat $in > $out")?;

        e.rule(
            "remove-imports",
            "sed '1,/component main/{/component main/!d}' $in > $out",
        )?;
        Ok(())
    });
    bld.op(
        "axi-wrapped",
        &[calyx_setup, wrapper_setup],
        calyx,
        calyx,
        |e, input, output| {
            // Generate the YXI file.
            //no extension
            let file_name = input
                .rsplit_once('/')
                .unwrap()
                .1
                .rsplit_once('.')
                .unwrap()
                .0;
            let tmp_yxi = format!("{}.yxi", file_name);

            //Get yxi file from main compute program.
            //TODO(nate): Can this use the `yxi` operation instead of hardcoding the build cmd calyx rule with arguments?
            e.build_cmd(&[&tmp_yxi], "calyx", &[input], &[])?;
            e.arg("backend", "yxi")?;

            // Generate the AXI wrapper.
            let refified_calyx = format!("refified_{}.futil", file_name);
            e.build_cmd(&[&refified_calyx], "calyx-pass", &[input], &[])?;
            e.arg("pass", "external-to-ref")?;

            let axi_wrapper = "axi_wrapper.futil";
            e.build_cmd(&[axi_wrapper], "gen-axi", &[&tmp_yxi], &[])?;

            // Generate no-imports version of the refified calyx.
            let no_imports_calyx = format!("no_imports_{}", refified_calyx);
            e.build_cmd(
                &[&no_imports_calyx],
                "remove-imports",
                &[&refified_calyx],
                &[],
            )?;

            // Combine the original Calyx and the wrapper.
            e.build_cmd(
                &[output],
                "combine",
                &[axi_wrapper, &no_imports_calyx],
                &[],
            )?;
            Ok(())
        },
    );
}
