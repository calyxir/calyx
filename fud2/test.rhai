
let dahlia2 = state("dahlia2", ["fuse"]);
let calyx2 = state("calyx2", ["futil"]);

fn dahlia_setup(e) {
   e.config_var("dahlia-exe", "dahlia");
   let test = e.config_val("sim.data");
   e.comment(test); 
   e.rule("dahlia2-to-calyx2", "$dahlia-exe -b calyx --lower -l error $in -o $out");
}

rule([dahlia_setup], dahlia2, calyx2, "dahlia2-to-calyx2");
