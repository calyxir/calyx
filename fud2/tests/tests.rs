use fud2::build_driver;
use fud_core::{
    config::default_config, exec::Request, run::Run, Driver, DriverBuilder,
};

fn test_driver() -> Driver {
    let mut bld = DriverBuilder::new("fud2");
    build_driver(&mut bld);
    bld.build()
}

fn request(
    driver: &Driver,
    start: &str,
    end: &str,
    through: &[&str],
) -> Request {
    fud_core::exec::Request {
        start_file: None,
        start_state: driver.get_state(start).unwrap(),
        end_file: None,
        end_state: driver.get_state(end).unwrap(),
        through: through.iter().map(|s| driver.get_op(s).unwrap()).collect(),
        workdir: ".".into(),
    }
}

fn emit_ninja(driver: &Driver, req: Request) -> String {
    let plan = driver.plan(req).unwrap();
    let config = default_config()
        .merge(("exe", "fud2"))
        .merge(("calyx.base", "/test/calyx"))
        .merge(("firrtl.exe", "/test/bin/firrtl"))
        .merge(("sim.data", "/test/data.json"))
        .merge(("xilinx.vivado", "/test/xilinx/vivado"))
        .merge(("xilinx.vitis", "/test/xilinx/vitis"))
        .merge(("xilinx.xrt", "/test/xilinx/xrt"))
        .merge(("dahlia", "/test/bin/dahlia"));
    let run = Run::with_config(driver, plan, config);
    let mut buf = vec![];
    run.emit(&mut buf).unwrap();
    String::from_utf8(buf).unwrap()
}

/// Get a human-readable description of a request.
fn req_desc(driver: &Driver, req: &Request) -> String {
    let mut desc = format!(
        "emit {} -> {}",
        driver.states[req.start_state].name, driver.states[req.end_state].name
    );
    if !req.through.is_empty() {
        desc.push_str(" through");
        for op in &req.through {
            desc.push(' ');
            desc.push_str(&driver.ops[*op].name);
        }
    }
    desc
}

/// Get a short string uniquely identifying a request.
fn req_slug(driver: &Driver, req: &Request) -> String {
    let mut desc = driver.states[req.start_state].name.to_string();
    for op in &req.through {
        desc.push('_');
        desc.push_str(&driver.ops[*op].name);
    }
    desc.push('_');
    desc.push_str(&driver.states[req.end_state].name);
    desc
}

fn test_emit(driver: &Driver, req: Request) {
    let desc = req_desc(driver, &req);
    let slug = req_slug(driver, &req);
    let ninja = emit_ninja(driver, req);
    insta::with_settings!({
        description => desc,
        omit_expression => true,
        snapshot_suffix => slug,
    }, {
        insta::assert_snapshot!(ninja);
    });
}

#[test]
fn calyx_to_verilog() {
    let driver = test_driver();
    test_emit(&driver, request(&driver, "calyx", "verilog", &[]));
}

#[test]
fn calyx_via_firrtl() {
    let driver = test_driver();
    test_emit(&driver, request(&driver, "calyx", "verilog", &["firrtl"]));
}

#[test]
fn sim_tests() {
    let driver = test_driver();
    for dest in &["dat", "vcd"] {
        for sim in &["icarus", "verilator"] {
            test_emit(&driver, request(&driver, "calyx", dest, &[sim]));
        }
    }
}

#[test]
fn cider_tests() {
    let driver = test_driver();
    test_emit(&driver, request(&driver, "calyx", "dat", &["interp"]));
    test_emit(&driver, request(&driver, "calyx", "debug", &[]));
}

#[test]
fn xrt_tests() {
    let driver = test_driver();
    test_emit(&driver, request(&driver, "calyx", "dat", &["xrt"]));
    test_emit(&driver, request(&driver, "calyx", "vcd", &["xrt-trace"]));
}

#[test]
fn frontend_tests() {
    let driver = test_driver();
    for frontend in &["dahlia", "mrxl"] {
        test_emit(&driver, request(&driver, frontend, "calyx", &[]));
    }
}
