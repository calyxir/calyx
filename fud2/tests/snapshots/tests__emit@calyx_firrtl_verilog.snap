---
source: fud2/tests/tests.rs
description: emit calyx -> verilog through firrtl
---
build-tool = fud2
rule get-rsrc
  command = $build-tool get-rsrc $out

# Calyx compiler
calyx-base = /test/calyx
calyx-exe = $calyx-base/target/debug/calyx
rule calyx
  command = $calyx-exe -l $calyx-base -b $backend $args $in > $out
rule calyx-pass
  command = $calyx-exe -l $calyx-base -p $pass $args $in > $out

# Firrtl to Verilog compiler
firrtl-exe = /test/bin/firrtl
rule firrtl
  command = $firrtl-exe -i $in -o $out -X sverilog
build primitives-for-firrtl.sv: get-rsrc
rule add-firrtl-prims
  command = cat primitives-for-firrtl.sv $in > $out

# build targets
build stdin.fir: calyx stdin
  backend = firrtl
build partial.sv: firrtl stdin.fir
build stdin.sv: add-firrtl-prims partial.sv | primitives-for-firrtl.sv

default stdin.sv
